src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,489 @@
+
+    @ExportMessage
+    protected RubyArray shift(RubyHash hash,
+            @Shared @Cached HashingNodes.ToHash hashFunction,
+            @Shared @Cached GetHashPosInIndexNode getHashPos,
+            @CachedLibrary("this") HashStoreLibrary self,
+            @Cached @Exclusive ConditionProfile noKvToShift) {
+        int lastKeyPos = getLastKeyPos();
+        if (noKvToShift.profile(lastKeyPos < 0)) {
+            return null;
+        }
+        Object key = kvStore[lastKeyPos];
+        int keyHash = hashFunction.execute(key, hash.compareByIdentity);
+        int hashPos = getHashPos.execute(keyHash, index);
+        Object val = deleteKvAndGetV(hash, hashPos, lastKeyPos);
+
+        return ArrayHelpers.createArray(
+                RubyContext.get(self),
+                RubyLanguage.get(self),
+                new Object[]{ key, val });
+    }
+
+    @ExportMessage
+    protected void rehash(RubyHash hash,
+            @Cached @Shared HashingNodes.ToHash hashFunction,
+            @Cached @Exclusive ConditionProfile slotUsed) {
+        for (int i = 0; i < index.length; i += 2) {
+            int kvOffset = index[i + 1];
+
+            if (slotUsed.profile(kvOffset > INDEX_SLOT_UNUSED_MARKER)) {
+                kvOffset--;
+
+                Object key = kvStore[kvOffset];
+                index[i] = hashFunction.execute(key, hash.compareByIdentity);
+            }
+        }
+    }
+
+    @ExportMessage
+    protected boolean verify(RubyHash hash) {
+        assert kvStoreInsertionPos > 0 && kvStoreInsertionPos < kvStore.length;
+        assert kvStoreInsertionPos % 2 == 0;
+
+        assert kvStoreInsertionPos >= hash.size; // because deletes only decrease hash.size
+        assert ((float) (2 * hash.size)) / index.length <= THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD;
+        return true;
+    }
+
+    private void insertIntoIndex(int keyHash, int kvPos) {
+        int pos = getIndexPosFromHash(keyHash, index.length);
+
+        boolean slotFilled = index[pos + 1] > INDEX_SLOT_UNUSED_MARKER;
+        while (slotFilled) {
+            pos = (pos + 2) & (index.length - 1);
+            slotFilled = index[pos + 1] > INDEX_SLOT_UNUSED_MARKER;
+        }
+
+        index[pos] = keyHash;
+        index[pos + 1] = kvPos;
+    }
+
+    private void insertIntoKv(Object key, Object value) {
+        kvStore[kvStoreInsertionPos] = key;
+        kvStore[kvStoreInsertionPos + 1] = value;
+        kvStoreInsertionPos += 2;
+    }
+
+    private void resizeIndexIfNeeded(int size) {
+        if (size < numSlotsForIndexRebuild) {
+            return;
+        }
+
+        int[] oldIndex = this.index;
+        this.index = new int[2 * index.length];
+
+        for (int i = 0; i < oldIndex.length; i += 2) {
+            int hash = oldIndex[i];
+            int kvPos = oldIndex[i + 1];
+
+            insertIntoIndex(hash, kvPos);
+        }
+        this.numSlotsForIndexRebuild = (int) (this.index.length / 2 * THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD);
```

Comment:
This is a lot of work so we want to move that logic behind a boundary and we want a CondiitonProfile on the check above

Reply:
Could you expand a bit on why we should move the logic behind a boundary ? My understanding is that boundaries are for things PE are bad at, like e.g. Recursion. Why would PE struggle with that allocation and loop ?

Reply:
Here it's just a judgement call. It's an operation that should happen infrequently. Due to its size, it'll likely compile to a large chunk of code and probably won't ever have the chance to pay for itself. Since the loop size isn't fixed, the compiler is limited in just how much it can really improve things here.

Reply:
See https://github.com/oracle/truffleruby/pull/3172/files#r1337242045